### DISCUSSION TITLE
Is it possible vsg viewer's TransferTask support partial update image?

### DISCUSSION THREAD

**Discussion by simwangrui** [vsg-dev/VulkanSceneGraph#743] [Q&A] [answered]
Hi,
I am using VSG  developing my terrain based on geometry clipmap . It caches terrain geometry in a set of nested regular grids, which are incrementally shifted as the viewer moves.This means that the image is partially updated as the viewer moves.

I noticed that vsg viewer's TransferTask can help transfer images in the background.My current understanding is that TransferTask can only update the image as a whole,is it possible vsg viewer's TransferTask support partial update image?

Thanks

best wishes

**Comment by robertosfield** [comment, marked as answer]
I have converted this questions to  Discussion Q&A as this is the appropriate place for such discussions.

TransferTask can only copy complete BufferInfo and ImageInfo but in the case of BufferInfo you can created your own BufferInfo that addresses a portion of the overall BufferInfo.  You can't presently do the same for ImageInfo as they are backed by a very specific type of object on the GPU and would need more elaborate mechanism.

One possible alternative would be to copy the data via BufferInfo and then copy this to the image, this would be too disimilar to the [vsgdynamictexture_cs](https://github.com/vsg-dev/vsgExamples/blob/master/examples/state/vsgdynamictexture_cs/vsgdynamictexture_cs.cpp) example does, where it reformats RGB data into RGBA that can be assigned to the vkImage.

**Comment by simwangrui** [comment]
Hi Robert,
Before I generated the clipmap WGS84 surface,I learned vsgdynamictexture_ cs example.Inspired by this example, I use compute shaders to generate surface vertices and normals.Surface vertices Stored in SSBO of  VK_ FORMAT_ R32G32B32A32_SINT format .now surface is working,I want to add elevation on the surface,using SSBO of VK_FORMAT_R16_SINT format.


The following is my idea, which has not been verified：
Because VK_ FORMAT_ R16_ SINT is an integer multiple of 4，so I may copy the 16bit elevation data to the staging buffer, and then transfer it to the GPU image through the vkCmdCopyBufferToImage command.I use the VkBufferImageCopy parameter to control the transfer of elevation data to the specified area of the image.imageOffset  controls the starting coordinates of image sub-regions and imageExtent  controls the size of the subarea.

typedef struct VkBufferImageCopy {
    VkDeviceSize            bufferOffset;
    uint32_t                    bufferRowLength;
    uint32_t                    bufferImageHeight;
    VkImageSubresourceLayers    imageSubresource;
    VkOffset3D                  imageOffset;
    VkExtent3D                  imageExtent;
} VkBufferImageCopy;

If this idea is feasible, can this function be implemented in TransferTask? Or I must implement this  myself？

Thanks.

**Reply by robertosfield** [reply]
The vsg::TransferTask API and implementation isn't set it stone so potentially we could extend it further, what you are aiming for is complex enough that I can point you in possible directions but not provide definitive answers and guidance.  I like everyone else am feeling my way through how mold a scene graph around Vulkan and modern graphics hardware.

As a next step I would suggest trying to do the copy in the compute stage as it's done in the vsgdyanmictexture_cs example.  The TransferTask is run before the compute shaders so should make the basic data available, you can then use the compute shader to transfer or simple put the copy at the beginning of the graphics command graph.

**Reply by simwangrui** [reply]
Hi,

Thank you for your detailed suggestions. I will try to do it according to your ideas.

Best wishes.
